### ISSUE TITLE
sub-document virtuals in nested arrays don't get attached when specified

### ISSUE THREAD

**Issue by iammola** [mongoosejs/mongoose-lean-virtuals#62]
### Do you want to request a _feature_ or report a _bug_?

Bug

### What is the current behaviour?

The virtuals of sub-documents in nested arrays aren't attached to the result when I specify the virtual paths `{ virtuals: [...] }`. It works fine when I lean the query with `{ virtuals: true }`

#### If the current behaviour is a bug, please provide the steps to reproduce.

```js
import mongoose from "mongoose";
import { mongooseLeanVirtuals } from "mongoose-lean-virtuals";

mongoose.plugin(mongooseLeanVirtuals);

const NameSchema = new mongoose.Schema({
  first: { type: String, required: true },
  last: String,
});

NameSchema.virtual("full").get(function () {
  return `${this.first} ${this.last ?? ""}`.trim();
});

const ChildSchema = new mongoose.Schema({
  age: { type: Number, required: true },
  name: { type: NameSchema, required: true },
}, { _id: false });

const ParentModel = mongoose.model(
  "Parent",
  new mongoose.Schema({
    name: { type: NameSchema, required: true },
    child: ChildSchema,
    nested: new mongoose.Schema({
      children: [ChildSchema],
    }),
  })
);

async function run() {
  await mongoose.connect("...");

  await ParentModel.create({
    name: { first: "Homer", last: "Simpson" },
    child: { age: 10, name: { first: "Bart", last: "Simpson" } },
    nested: {
      children: [
        { age: 6, name: { first: "Lisa", last: "Simpson" } },
        { age: 3, name: { first: "Baby" } },
      ],
    },
  });

  const result = await ParentModel.find({})
    .populate("child")
    .populate("nested.children")
    .lean({
      virtuals: ["name.full", "child.name.full", "children.name.full"],
    });

  assert.equal(result.name.full, "Homer Simpson"); // Pass
  assert.equal(result.child.name.full, "Bart Simpson"); // Pass
  assert.equal(result.children[0].name.full, "Lisa Simpson"); // Fail
  assert.equal(result.children[1].name.full, "Baby"); // Fail
}

run().catch(console.error)
```

#### What is the expected behaviour?

For the virtuals of the sub-documents in nested arrays to be attached.

#### What are the versions of Node.js, mongoose-lean-getters, and Mongoose are you are using? Note that "latest" is not a version.

| Package               | Version |
| --------------------- | ------- |
| mongoose              | 6.4.3   |
| mongoose-lean-getters | N/A     |
| Node.js               | 18.6.0  |

**Comment by vkarpov15** [comment]
You need to specify `nested.children.name.full`. The script you provided fails, but the below script works:

```javascript
const mongoose = require('mongoose');
const mongooseLeanVirtuals = require('mongoose-lean-virtuals');
const assert = require('assert');

mongoose.plugin(mongooseLeanVirtuals);

const NameSchema = new mongoose.Schema({
  first: { type: String, required: true },
  last: String,
});

NameSchema.virtual("full").get(function () {
  return `${this.first} ${this.last ?? ""}`.trim();
});

const ChildSchema = new mongoose.Schema({
  age: { type: Number, required: true },
  name: { type: NameSchema, required: true },
}, { _id: false });

const ParentModel = mongoose.model(
  "Parent",
  new mongoose.Schema({
    name: { type: NameSchema, required: true },
    child: ChildSchema,
    nested: new mongoose.Schema({
      children: [ChildSchema],
    }),
  })
);

async function run() {
  await mongoose.connect("mongodb://localhost:27017/test");
  await mongoose.connection.dropDatabase();

  const { _id } = await ParentModel.create({
    name: { first: "Homer", last: "Simpson" },
    child: { age: 10, name: { first: "Bart", last: "Simpson" } },
    nested: {
      children: [
        { age: 6, name: { first: "Lisa", last: "Simpson" } },
        { age: 3, name: { first: "Baby" } },
      ],
    },
  });

  const result = await ParentModel.findById(_id)
    .populate("child")
    .populate("nested.children")
    .lean({
      virtuals: ["name.full", "child.name.full", "nested.children.name.full"], // <-- note the 'nested.'
    });

  assert.equal(result.name.full, "Homer Simpson"); // Pass
  assert.equal(result.child.name.full, "Bart Simpson"); // Pass
  assert.equal(result.nested.children[0].name.full, "Lisa Simpson"); // Pass
  assert.equal(result.nested.children[1].name.full, "Baby"); // Pass
  console.log('Done');
}

run().catch(console.error)
```

Also works fine if you remove the 'nested'

**Comment by iammola** [comment]
Hey @vkarpov15, forgive my invalid script. I'll try to explain my actual use case and give an excuse for why my script had errors.

I have a model with 2 discriminator schemas applied to it. The `ParentModel.child` and `ParentModel.nested...` was meant to show the field I'm trying to populate between them. They both point to the same model it just doesn't work with virtuals on the array of sub-documents.

So I created a repro script to use here with populating the fields, and it didn't work. So I tried to check if it affected other schema definitions as well. That's why my original script has `.populate("...")` calls when the fields don't need it and the values aren't correctly checked.

This is another script that I'm pretty sure will show my issue. Thanks again for your putting up with me and all your efforts across all the mongoose packages.

```ts
import assert from "node:assert";

import mongoose from "mongoose";
import { mongooseLeanVirtuals } from "mongoose-lean-virtuals";

async function run() {
  mongoose.plugin(mongooseLeanVirtuals);
  await mongoose.connect("...");

  const NameSchema = new mongoose.Schema({
    first: { type: String, required: true },
    last: String,
  });

  NameSchema.virtual("full").get(function () {
    return `${this.first} ${this.last ?? ""}`.trim();
  });

  const ChildModel = mongoose.model(
    "Child",
    new mongoose.Schema({
      age: { type: Number, required: true },
      name: { type: NameSchema, required: true },
    })
  );

  const ParentModel = mongoose.model(
    "Parent",
    new mongoose.Schema({
      name: { type: NameSchema, required: true },
      child: { type: mongoose.Types.ObjectId, ref: "Child" },
      nested: {
        type: [new mongoose.Schema({ item: { type: mongoose.Types.ObjectId, ref: "Child" } })],
      },
    })
  );

  const [child_1, child_2, child_3] = await ChildModel.create([
    { age: 10, name: { first: "Bart", last: "Simpson" } },
    { age: 6, name: { first: "Lisa", last: "Simpson" } },
    { age: 3, name: { first: "Baby" } },
  ]);

  await ParentModel.create({
    name: { first: "Homer", last: "Simpson" },
    child: child_1._id,
    nested: [{ item: child_3._id }, { item: child_2._id }],
  });

  const result = await ParentModel.findOne({})
    .populate("child")
    .populate("nested.item")
    .orFail(new Error("DOC!"))
    .lean({ virtuals: ["name.full", "child.name.full", "nested.item.name.full"] });

  assert.equal(result.name.full, "Homer Simpson"); // Pass
  assert.equal(result.child?.name.full, "Bart Simpson"); // Pass
  assert.equal(result.nested[0].item?.name.full, "Baby"); // Fail
  assert.equal(result.nested[1].item?.name.full, "Lisa Simpson"); // Fail
}

run();
```

**Comment by iammola** [comment]
Also, did you mean to close #36? The PR you linked didn't have an effect because you used the `Fix` keyword on #37

**Comment by vkarpov15** [comment]
We took a closer look and this proves to be tricky to implement in general with how Mongoose populate works. Without hooking more closely into Mongoose populate, handling cases like discriminators isn't really feasible. However, there is a simple workaround:

```javascript
  const result = await ParentModel.findOne({})
    .populate("child")
    .populate({ path: "nested.item", options: { lean: { virtuals: ['name.full'] } } }) // <-- add the `name.full` virtual here
    .orFail(new Error("DOC!"))
    .lean({ virtuals: ["name.full", "child.name.full"] }); // <-- instead of here
```
